src/datumaro/components/prune.py

```diff
@@ -0,0 +1,322 @@
+
+        if self._cluster_method == "random":
+            self._item_list = list(self._dataset)
+        else:
+            datasets_to_infer = select_uninferenced_dataset(self._dataset)
+            datasets = self._compute_hash_key([self._dataset], [datasets_to_infer])[0]
+
+            for category in datasets.categories().values():
+                if isinstance(category, LabelCategories):
+                    self._num_centers = len(category._indices.keys())
+
+            for item in datasets:
+                for annotation in item.annotations:
+                    if isinstance(annotation, Label):
+                        self._labels.append(annotation.label)
+                    if isinstance(annotation, HashKey):
+                        hash_key = annotation.hash_key[0]
+                        if self._hash_type == "txt":
+                            inputs = category_dict.get(item.annotations[0].label)
+                            if isinstance(inputs, List):
+                                inputs = " ".join(inputs)
+                            hash_key_txt = self._text_model.infer_text(inputs)
+                            hash_key = np.concatenate([hash_key, hash_key_txt])
+                        hash_key = np.unpackbits(hash_key, axis=-1)
+                        if self._database_keys is None:
+                            self._database_keys = hash_key.reshape(1, -1)
+                        else:
+                            self._database_keys = np.concatenate(
+                                (self._database_keys, hash_key.reshape(1, -1)), axis=0
+                            )
+                self._item_list.append(item)
+
+    @property
+    def model(self):
+        if self._model is None:
+            self._model = ExplorerLauncher(model_name="clip_visual_ViT-B_32")
+        return self._model
+
+    @property
+    def text_model(self):
+        if self._text_model is None:
+            self._text_model = ExplorerLauncher(model_name="clip_text_ViT-B_32")
+        return self._text_model
+
+    def prompting(self):
+        category_dict = {}
+        detected_format = self._dataset.format
+        template = format_templates.get(detected_format, templates)
+        for label, indice in list(self._dataset.categories().values())[0]._indices.items():
+            category_dict[indice] = [temp.format(label) for temp in template]
+        return category_dict
+
+    def _compute_hash_key(self, datasets, datasets_to_infer):
+        for dataset_to_infer in datasets_to_infer:
+            if dataset_to_infer:
+                dataset_to_infer.run_model(self.model, append_annotation=True)
+        for dataset, dataset_to_infer in zip(datasets, datasets_to_infer):
+            updated_items = [
+                dataset.get(item.id, item.subset).wrap(annotations=item.annotations)
+                for item in dataset_to_infer
+            ]
+            dataset.update(updated_items)
+        return datasets
```

Comment:
Please make a common ancestor for `Explorer` and `Pruner` to remove duplicate functions.

Reply:
Make `HashInference` for `Explorer` and `Pruner`. If it is not proper, please let me know.